feat(coding-guidelines): add guideline for ros node directory structure (#224)
autowarefoundation/autoware-documentation: docs/contributing/coding-guidelines/ros-nodes/directory-structure.md

# Directory structure

## C++ package

```txt
<package-name>
├─ config
├─ include
│   └─ <package-name>
├─ launch
├─ src
├─ test
├─ package.xml
└─ CMakeLists.txt
```

### config directory

Place configuration files such as node parameters.

### include directory

Place header files exposed to other packages under a directory with the package name. Do not place files directly under include directory.

> This is mostly library headers and such. Note that many headers do not need to be placed here.

Reviewer: I agree :+

Author: Done https://github.com/autowarefoundation/autoware-documentation/pull/224/commits/763108b5b2b7440e9474bed2e33ecab472b4cc75.

### launch directory

Place launch files.

### src directory

Place source and private header files.

### test directory

Place source files for testing.

## Python package

!!! warning

>     Under Construction

Reviewer 2: Could you add content under this warning or remove the warning?

Author: Done https://github.com/autowarefoundation/autoware-documentation/pull/224/commits/763108b5b2b7440e9474bed2e33ecab472b4cc75.

!!! warning

    Under Construction

## C++ package

```txt

> <package-name>
> ├─ config
> ├─ include
> │   └─ <package-name>
> ├─ launch
> ├─ src
> ├─ test
> ├─ package.xml
> └─ CMakeLists.txt

Reviewer 2: How about changing like this?
- Use snake_case.
- Add some example files.

```suggestion
<package_name>
├─ config
│   └─ <package_name>
├─ include
│   └─ <package_name>
│      └─ foo_public.hpp
├─ launch
│   ├─ foo.launch.xml
│   └─ foo.launch.py
├─ src
│   ├─ foo_node.cpp
│   ├─ foo_node.hpp
│   └─ foo_private.hpp
├─ test
│   └─ test_foo.cpp
├─ package.xml
└─ CMakeLists.txt
```

!!! warning

    Under Construction

## C++ package

```txt
<package_name>
├─ config

> │   └─ <package_name>

Reviewer 2: Sorry, there was a mistake.

```suggestion
│   └─ foo.param.yaml
```

> │   └─ foo.param.yaml

Reviewer 2: Considering https://github.com/autowarefoundation/autoware-documentation/pull/224/files#r1005413621, should we write like this? :thinking: 

```suggestion
│   ├─ foo_ros.param.yaml
│   └─ foo_non_ros.yaml
```

├─ include
│   └─ <package_name>
│      └─ foo_public.hpp
├─ launch
│   ├─ foo.launch.xml
│   └─ foo.launch.py
├─ src
│   ├─ foo_node.cpp
│   ├─ foo_node.hpp
│   └─ foo_private.hpp
├─ test
│   └─ test_foo.cpp
├─ package.xml
└─ CMakeLists.txt
```

### config directory

> Place configuration files such as node parameters.

Reviewer 2: How about adding this?

```suggestion
Place configuration files such as node parameters.

For ROS parameters, use the extension `.param.yaml`.
For non-ROS parameters, use the extension `.yaml`.

Rationale: Since ROS parameters files are type-sensitive, they should not be the target of some code formatters and linters. In order to distinguish the file type, we use different file extensions.
```

!!! warning

    Under Construction

## C++ package

```txt
<package_name>
├─ config
│   ├─ foo_ros.param.yaml
│   └─ foo_non_ros.yaml
├─ include
│   └─ <package_name>
│      └─ foo_public.hpp
├─ launch
│   ├─ foo.launch.xml
│   └─ foo.launch.py
├─ src
│   ├─ foo_node.cpp
│   ├─ foo_node.hpp
│   └─ foo_private.hpp
├─ test
│   └─ test_foo.cpp
├─ package.xml
└─ CMakeLists.txt
```

### config directory

Place configuration files such as node parameters.

For ROS parameters, use the extension `.param.yaml`.
For non-ROS parameters, use the extension `.yaml`.

Rationale: Since ROS parameters files are type-sensitive, they should not be the target of some code formatters and linters. In order to distinguish the file type, we use different file extensions.

### include directory

> Place header files exposed to other packages under a directory with the package name. Do not place files directly under include directory.
> This is mostly library headers and such. Note that many headers do not need to be placed here.

Reviewer 2: I personally feel this is easier to understand. (But I'm not sure.)

```suggestion
Place header files exposed to other packages. Do not place files directly under the `include` directory, but place files under the directory with the package name.
This directory is used for mostly library headers. Note that many headers do not need to be placed here. It is enough to place the headers under the `src` directory.
```

Reference: <https://docs.ros.org/en/rolling/How-To-Guides/Ament-CMake-Documentation.html#adding-files-and-headers>

### launch directory

> Place launch files.

Reviewer 2: How about this?

```suggestion
Place `.launch.xml` and `.launxh.py`.
```
Or,

```suggestion
Place launch files (`.launch.xml` and `.launxh.py`).
```

### src directory

> Place source and private header files.

Reviewer 2: To be specific.

```suggestion
Place source files and private header files.
```

> Place launch files (`.launch.xml` and `.launxh.py`).

Reviewer 2: I'm so sorry.

```suggestion
Place launch files (`.launch.xml` and `.launch.py`).
```
